asmcomp/power/selection.ml

```diff
@@ -43,13 +43,28 @@ let rec select_addr = function
   | exp ->
       (Alinear exp, 0, Debuginfo.none)
 
+let is_immediate n = n <= 0x7FFF && n >= -0x8000
+let is_immediate_logical n = n <= 0xFFFF && n >= 0
+
 (* Instruction selection *)
 
 class selector = object (self)
 
 inherit Selectgen.selector_generic as super
 
-method is_immediate n = (n <= 32767) && (n >= -32768)
+method is_immediate_test cmp n =
+  match cmp with
+  | Isigned _ -> is_immediate n
+  | Iunsigned _ -> is_immediate_logical n
+
+method is_immediate op n =
+  match op with
+  | Iadd | Isub | Imul -> is_immediate n
+  | Imulh -> false  (* no immediate operands for multiply high *)
+  | Iand | Ior | Ixor -> is_immediate_logical n
+  | Icomp c -> self#is_immediate_test c n
+  | Icheckbound _ -> is_immediate n
+  | _ -> assert false
```

Comment:
None of the methods in $ARCH/selection.ml enumerates all cases explicitly: they all have a final catch-all case (which, generally, calls the method of the superclass).  That's in the spirit of this code: common behavior is implemented in selectgen.ml, and each $ARCH/selection.ml adds behavior only for the constructs that need special handling.

So, enumerating all cases here is not in the spirit of this code, I'm sorry.  But `_ -> false` could be a more appropriate default than `_ -> assert false`, just because it's always safe to not recognize an immediate argument.  Or `_ -> super#is_immediate c n` if we really want to be in the spirit.  This could also avoid special-casing shifts and divisions, as you remark.  Let me come back with an alternate proposal.

Reply:
Commit f8eed137c reimplements the `is_immediate` method in the spirit I was trying to describe.  There's a fall-back case `_ -> super#is_immediate c n` that invokes a default `is_immediate` method that deals with shifts and returns `false` otherwise.  Each port lists the Mach operations for which it has immediate forms, and none other.  The default behavior is to return `false`, which may be suboptimal (a supported immediate form is not recognized) but safe.

Reply:
This is nice, thanks!